caffe2/core/tensor.h

```diff
@@ -608,27 +638,36 @@ class CAFFE2_API TensorImpl : public c10::intrusive_ptr_target {
            storage_.capacity() >= numel_ * storage_.itemsize())) {
         return storage_.data();
       }
+      const at::Allocator* allocator = storage_.allocator();
+      CAFFE_ENFORCE(allocator == nullptr);
       if (meta.ctor()) {
         // For types that need placement new, we will call it, as well as
         // making sure that when the data is freed, it calls the right
         // destruction procedure.
         auto size = numel_;
         auto dtor = storage_.dtype().dtor();
+        void* ptr;
+        at::DeleterFnPtr deleter;
         auto ptr_and_deleter =
             GetStaticContext()->New(numel_ * storage_.itemsize());
-        auto deleter = ptr_and_deleter.second;
-        storage_.data_ptr().reset(
-            ptr_and_deleter.first, [size, dtor, deleter](void* ptr) -> void {
-              dtor(ptr, size);
-              deleter(ptr);
-            });
+        ptr = ptr_and_deleter.first;
+        deleter = ptr_and_deleter.second;
+        storage_.set_data_ptr(at::InefficientStdFunctionContext::makeDataPtr(
```

Comment:
It might be worth making a context specifically for this use-case. It will save you an extra indirection, assuming that `std::function` isn't storing the closed over variables inline (unlikely), and will also you save you from any `std::function` performance stupidity.